daml-lf/engine/src/main/scala/com/digitalasset/daml/lf/engine/preprocessing/ValueTranslator.scala

```diff
@@ -196,94 +199,109 @@ private[lf] final class ValueTranslator(
                 val targetFieldsAndTypes = lookupResult.dataRecord.fields
                 val subst = lookupResult.subst(tyArgs)
 
+                def addMissingField(lbl: Name, ty: Type): (Option[Name], Value) =
+                  ty match {
+                    // If missing field is optional, fill it with None
+                    case TApp(TBuiltin(BTOptional), _) => (Some(lbl), Value.ValueOptional(None))
+                    // Else, throw error
+                    case _ =>
+                      typeError(
+                        s"Missing non-optional field $lbl, cannot upgrade non-optional fields."
+                      )
+                  }
+
                 if (config.enableUpgrade) {
-                  // We handle the (non-)upgrade cases for record separately
-                  // In case of upgrade, we require the field are record.
-
-                  // This code implements the compatibility transformation used for up/down-grading
-                  // And handles the cases:
-                  // - UPGRADE:   numT > numS : creates a None for each missing fields.
-                  // - DOWNGRADE: numS > numT : drops each extra field, ensuring it is None.
-                  //
-                  // When numS == numT, we wont hit the code marked either as UPGRADE or DOWNGRADE,
-                  // although it is still possible that the source and target types are different,
-                  // but since we don't consult the source type (may be unavailable), we wont know.
-
-                  val numS: Int = sourceElements.length
-                  val numT: Int = targetFieldsAndTypes.length
-
-                  // traverse the sourceElements, "get"ing the corresponding target type
-                  // when there is no corresponding type, we must be downgrading, and so we insist the value is None
-                  val values0: List[SValue] =
-                    sourceElements.toSeq.view.zipWithIndex.flatMap { case ((optName, v), i) =>
-                      targetFieldsAndTypes.get(i) match {
-                        case Some((targetField, targetFieldType)) =>
-                          optName match {
-                            case None => ()
-                            case Some(sourceField) =>
-                              // value is not normalized; check field names match
-                              if (sourceField != targetField)
+                  val oLabeledFlds =
+                    labeledRecordToMap(sourceElements)
+                      .fold(typeError, identity _)
+                      .filter(Function.const(config.allowFieldReordering))
```

Comment:
I needed to comment out the `.filter` to get the tests to pass. An alternative is to ensure that the flag is set to `allowFieldReordering` flag is set to true whenever `enableUpgrade` is true.

Reply:
Yeah the latter is what should happen, I'm guessing we're constructing the upgrade translator config explicitly somewhere